src/Evolution/Systems/RelativisticEuler/Valencia/FixConservatives.cpp

```diff
@@ -0,0 +1,139 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/RelativisticEuler/Valencia/FixConservatives.hpp"
+
+#include <cmath>
+#include <cstddef>
+#include <ostream>
+#include <pup.h>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace RelativisticEuler {
+namespace Valencia {
+
+template <size_t Dim>
+FixConservatives<Dim>::FixConservatives(
+    const double minimum_rest_mass_density_times_lorentz_factor,
+    const double rest_mass_density_times_lorentz_factor_cutoff,
+    const double safety_factor_for_momentum_density,
+    const OptionContext& context)
+    : minimum_rest_mass_density_times_lorentz_factor_(
+          minimum_rest_mass_density_times_lorentz_factor),
+      rest_mass_density_times_lorentz_factor_cutoff_(
+          rest_mass_density_times_lorentz_factor_cutoff),
+      one_minus_safety_factor_for_momentum_density_(
+          1.0 - safety_factor_for_momentum_density) {
+  if (minimum_rest_mass_density_times_lorentz_factor_ >
+      rest_mass_density_times_lorentz_factor_cutoff_) {
+    PARSE_ERROR(context,
+                "The minimum value of the rest mass density times the Lorentz "
+                "factor must be less than or equal to the cutoff value of D.\n"
+                "Values given: D_min = "
+                    << minimum_rest_mass_density_times_lorentz_factor_
+                    << ", D_cutoff = "
+                    << rest_mass_density_times_lorentz_factor_cutoff_ << ".");
+  }
+}
+
+template <size_t Dim>
+void FixConservatives<Dim>::pup(PUP::er& p) noexcept {
+  p | minimum_rest_mass_density_times_lorentz_factor_;
+  p | rest_mass_density_times_lorentz_factor_cutoff_;
+  p | one_minus_safety_factor_for_momentum_density_;
+}
+
+template <size_t Dim>
+void FixConservatives<Dim>::operator()(
+    const gsl::not_null<Scalar<DataVector>*> tilde_d,
+    const gsl::not_null<Scalar<DataVector>*> tilde_tau,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*> tilde_s,
+    const tnsr::II<DataVector, Dim, Frame::Inertial>& inv_spatial_metric,
+    const Scalar<DataVector>& sqrt_det_spatial_metric) const noexcept {
+  const size_t number_of_points = get(sqrt_det_spatial_metric).size();
+  Variables<tmpl::list<::Tags::TempScalar<0>, ::Tags::TempScalar<1>>>
+      temp_buffer(number_of_points);
+
+  // rest mass density times Lorentz factor
+  DataVector& rho_w = get(get<::Tags::TempScalar<0>>(temp_buffer));
```

Comment:
This is the variable D, and you use D in the documentation. Why not use the consistent `d` in the code?

Reply:
I don't think we like single-character identifiers. However, I added a comment clarifying that this is indeed D.